sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp

```diff
@@ -1063,43 +1089,46 @@ pi_result piMemRelease(pi_mem Mem) {
   }
 
   if (--(Mem->RefCount) == 0) {
+    int Status = cm_support::CM_FAILURE;
+    std::lock_guard<std::mutex> Lock(Mem->Context->CmSurfaceManageLock);
+
     if (Mem->getMemType() == PI_MEM_TYPE_BUFFER) {
       _pi_buffer *PiBuf = static_cast<_pi_buffer *>(Mem);
-      // TODO implement libCM API failure logging mechanism, so that these
-      // failures are clearly distinguishable from other EMU plugin failures.
-      int Result =
-          Mem->Context->Device->CmDevicePtr->DestroySurface(PiBuf->CmBufferPtr);
-
-      if (Result != cm_support::CM_SUCCESS) {
-        return PI_INVALID_MEM_OBJECT;
+      if (Mem->isHostProvidedMem()) {
+        Status = Mem->Context->Device->CmDevicePtr->DestroyBufferUP(
+            PiBuf->BufferPtr.HostPtr);
+      } else {
+        Status = Mem->Context->Device->CmDevicePtr->DestroySurface(
+            PiBuf->BufferPtr.GPUPtr);
       }
     } else if (Mem->getMemType() == PI_MEM_TYPE_IMAGE2D) {
       _pi_image *PiImg = static_cast<_pi_image *>(Mem);
-      int Result = Mem->Context->Device->CmDevicePtr->DestroySurface(
-          PiImg->CmSurfacePtr);
-      if (Result != cm_support::CM_SUCCESS) {
-        return PI_INVALID_MEM_OBJECT;
+      if (Mem->isHostProvidedMem()) {
+        Status = Mem->Context->Device->CmDevicePtr->DestroySurface2DUP(
+            PiImg->ImagePtr.HostPtr);
+      } else {
+        Status = Mem->Context->Device->CmDevicePtr->DestroySurface(
+            PiImg->ImagePtr.GPUPtr);
       }
-    } else {
-      return PI_INVALID_MEM_OBJECT;
     }
 
-    // Removing Surface-map entry
-    const std::lock_guard<sycl::detail::SpinLock> Lock{*PiESimdSurfaceMapLock};
-    auto MapEntryIt = PiESimdSurfaceMap->find(Mem->SurfaceIndex);
-    if (MapEntryIt != PiESimdSurfaceMap->end()) {
-      PiESimdSurfaceMap->erase(MapEntryIt);
-    } else {
-      if (PrintPiTrace) {
-        std::cerr << "Failure from CM-managed buffer/image deletion"
-                  << std::endl;
-      }
+    if (Status != cm_support::CM_SUCCESS) {
       return PI_INVALID_MEM_OBJECT;
     }
+  }
 
-    delete Mem;
+  // Removing Surface-map entry
```

Comment:
Lock is missing here to guard access to PiESimdSurfaceMap. This is a bug, so please fix